### PR TITLE
UI: refactor model related functions

### PR DESCRIPTION
**Pull request by deanlee**
- follow capnp best practices,make it run faster.
- Use template to merge functions `update_track_data`&`update_line_data` into one.
- Complete TODO: check that this doesn't overflow max vertex buffer.
- Use `scene->lane_line_probs[i]` instead of `scene->model.getLaneLineProbs()[ll_idx]` (looks like a bug)
- Remove function `ui_draw_track`

Next time, we can  move these functions to background thread to populate data.

### CODE REVIEW

File: selfdrive/ui/paint.cc [hunk 1]
```diff
@@ -208,49 +176,55 @@ static void draw_frame(UIState *s) {
...
-static void update_line_data(UIState *s, const cereal::ModelDataV2::XYZTData::Reader &line, float off, line_vertices_data *pvd, float max_distance) {
-  // TODO check that this doesn't overflow max vertex buffer
+template <class T>
+static void update_line_data(const UIState *s, const cereal::ModelDataV2::XYZTData::Reader &line,
+                             float y_off, float z_off, T *pvd, float max_distance) {
+  const auto line_x = line.getX(), line_y = line.getY(), line_z = line.getZ();
   int max_idx = -1;
   vertex_data *v = &pvd->v[0];
   const float margin = 500.0f;
   for (int i = 0; ((i < TRAJECTORY_SIZE) and (line.getX()[i] < fmax(MIN_DRAW_DISTANCE, max_distance))); i++) {
```
Comment (pd0wm) [inline selfdrive/ui/paint.cc:186]:
```suggestion
  for (int i = 0; ((i < TRAJECTORY_SIZE) and line_x[i] < fmax(MIN_DRAW_DISTANCE, max_distance))); i++) {
```

File: selfdrive/ui/paint.cc [hunk 2]
```diff
@@ -208,49 +176,55 @@ static void draw_frame(UIState *s) {
...
   }
 
   // paint road edges
-  line_vertices_data *pvd_re = &s->road_edge_vertices[0];
-  for (int re_idx = 0; re_idx < 2; re_idx++) {
-    if (s->sm->updated("modelV2")) {
-      update_line_data(s, scene->model.getRoadEdges()[re_idx], 0.025, pvd_re + re_idx, scene->max_distance);
+  for (int i = 0; i < std::size(s->road_edge_vertices); i++) {
+    if (model_updated) {
+      update_line_data(s, scene->model.getRoadEdges()[i], 0.025, 1.22, &s->road_edge_vertices[i], scene->max_distance);
```
Comment (pd0wm) [inline selfdrive/ui/paint.cc:212]: get road edges outside of for loop to follow capnp best practicies? Same for lane lines above.

**Comment by pd0wm** [comment]
Why is lane_line_probs stored on the scene. Shouldn't we just pull it out of the model packet everywhere?

File: selfdrive/ui/paint.cc [hunk 3]
```diff
@@ -208,49 +176,55 @@ static void draw_frame(UIState *s) {
...
+template <class T>
+static void update_line_data(const UIState *s, const cereal::ModelDataV2::XYZTData::Reader &line,
+                             float y_off, float z_off, T *pvd, float max_distance) {
+  const auto line_x = line.getX(), line_y = line.getY(), line_z = line.getZ();
   int max_idx = -1;
   vertex_data *v = &pvd->v[0];
   const float margin = 500.0f;
-  for (int i = 0; ((i < TRAJECTORY_SIZE) and (line.getX()[i] < fmax(MIN_DRAW_DISTANCE, max_distance))); i++) {
-    v += car_space_to_full_frame(s, line.getX()[i], -line.getY()[i] - off, -line.getZ()[i] + 1.22, &v->x, &v->y, margin);
+  for (int i = 0; ((i < TRAJECTORY_SIZE) and line_x[i] < fmax(MIN_DRAW_DISTANCE, max_distance))); i++) {
```
Comment (pd0wm) [inline selfdrive/ui/paint.cc:186]:
```suggestion
  for (int i = 0; ((i < TRAJECTORY_SIZE) and (line_x[i] < fmax(MIN_DRAW_DISTANCE, max_distance))); i++) {
```

Comment (pd0wm) [reply selfdrive/ui/paint.cc]: My bad...

**Comment by deanlee** [comment]
yes,we can safely remove lane_line_probs & road_edge_stds from scene

**Comment by deanlee** [comment]
I'm going to move them to ui.cc

**Comment by deanlee** [comment]
moved model related variables to UIScene

**Review by Torresangl925 — APPROVED** [review]

**Comment by pd0wm** [comment]
Merged!

**Pull request closed without merging**